examples/onnx/cnnonnx2singa.py

```diff
@@ -0,0 +1,123 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import numpy as np
+import argparse
+import os
+
+from singa import device
+from singa import tensor
+from singa import autograd
+from singa import opt
+from singa import sonnx
+import onnx
+
+
+def load_data(path):
+    f = np.load(path)
+    x_train, y_train = f['x_train'], f['y_train']
+    x_test, y_test = f['x_test'], f['y_test']
+    f.close()
+    return (x_train, y_train), (x_test, y_test)
+
+
+def to_categorical(y, num_classes):
+    '''
+    Converts a class vector (integers) to binary class matrix.
+    Args
+        y: class vector to be converted into a matrix
+            (integers from 0 to num_classes).
+        num_classes: total number of classes.
+    Return
+        A binary matrix representation of the input.
+    '''
+    y = np.array(y, dtype='int')
+    n = y.shape[0]
+    categorical = np.zeros((n, num_classes))
+    categorical[np.arange(n), y] = 1
+    categorical = categorical.astype(np.float32)
+    return categorical
+
+
+def preprocess(data):
+    data = data.astype(np.float32)
+    data /= 255
+    data = np.expand_dims(data, axis=1)
+    return data
+
+
+def accuracy(pred, target):
+    y = np.argmax(pred, axis=1)
+    t = np.argmax(target, axis=1)
+    a = y == t
+    return np.array(a, 'int').sum() / float(len(t))
+
+
+if __name__ == '__main__':
+
+
+    file_path = 'mnist.npz'
+    use_cpu = 'store_true'
+
+    print(file_path)
+    assert os.path.exists(file_path), \
+        'Pls download the MNIST dataset from https://s3.amazonaws.com/img-datasets/mnist.npz'
+    use_cpu=False
+    if use_cpu:
+        print('Using CPU')
+        dev = device.get_default_device()
+    else:
+        print('Using GPU')
+        dev = device.create_cuda_gpu()
+
+    train, test = load_data(file_path)
+    print(train[0].shape)
+    batch_number = 600
+    num_classes = 10
+    epochs = 1
+
+    sgd = opt.SGD(lr=0.01)
+
+    x_train = preprocess(train[0])
+    y_train = to_categorical(train[1], num_classes)
+
+    x_test = preprocess(test[0])
+    y_test = to_categorical(test[1], num_classes)
+    print('the shape of training data is', x_train.shape)
+    print('the shape of training label is', y_train.shape)
+    print('the shape of testing data is', x_test.shape)
+    print('the shape of testing label is', y_test.shape)
+
+    # operations initialization
+    conv1 = autograd.Conv2d(1, 32, 3, padding=1)
+    conv21 = autograd.Conv2d(32, 16, 3, padding=1)
+    conv22 = autograd.Conv2d(32, 16, 3, padding=1)
+    linear = autograd.Linear(32 * 28 * 28, 10)
+
+    model = sonnx.from_onnx_model('cnn.onnx')
+    print('finish init')
+    autograd.training = True
+    # training process
```

Comment:
Why there is another training process here? The model in the file `cnn.onnx` was already trained in `mnist_cnn.py`.

It is expected to use the trained model here by passing an input image and predicting the output without training again.

Reply:
I think that we can provide one more options to continue to train the model such as fine tune?

Reply:
we can load onnx model directly for inference.
we can also load it to continue the training.

Reply:
@nudles Yes, but I do not see an example for loading the model for inference and I am worried that the model parameters are not saved and loaded correctly. This is why I propose to show that inference is working without re-training.